File: app/src/main/java/de/wollis_page/gibsonos/activity/MainActivity.java
package de.wollis_page.gibsonos.activity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.List;

import androidx.annotation.Nullable;
import de.wollis_page.gibsonos.R;
import de.wollis_page.gibsonos.activity.base.AppCompatListActivity;
import de.wollis_page.gibsonos.adapter.AccountAdapter;
import de.wollis_page.gibsonos.model.Account;

public class MainActivity extends AppCompatListActivity
{
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_main);
        super.onCreate(savedInstanceState);
        this.navigationView.setNavigationItemSelectedListener(this);
        this.loadList();

        FloatingActionButton addButton = findViewById(R.id.addButton);
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivityForResult(new Intent(getApplicationContext(), LoginActivity.class), 100);
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == 100 && resultCode == RESULT_OK) {
            this.loadList();
        }
    }

    private void loadList() {
        final List<Account> accounts = this.application.getAccountModels();

        if (accounts.size() == 0) {
            this.startActivityForResult(new Intent(this.getApplicationContext(), LoginActivity.class), 100);
        }

        AccountAdapter accountAdapter = new AccountAdapter(this.getApplicationContext(), accounts);
        this.setListAdapter(accountAdapter);

        this.getListView().setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                runActivity(DesktopActivity.class, accounts.get(i));
            }
        });
    }
}

File: app/src/main/java/de/wollis_page/gibsonos/activity/MainActivity.kt
package de.wollis_page.gibsonos.activity

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import de.wollis_page.gibsonos.R
import de.wollis_page.gibsonos.dto.ListItemInterface
import de.wollis_page.gibsonos.model.Account
import de.wollis_page.gibsonos.module.core.desktop.activity.IndexActivity
import de.wollis_page.gibsonos.module.core.task.UserTask
import de.wollis_page.gibsonos.module.core.user.activity.LoginActivity
import de.wollis_page.gibsonos.service.ActivityLauncherService
import de.wollis_page.gibsonos.service.AppIntentExtraService
import de.wollis_page.gibsonos.dto.Account as AccountDto

class MainActivity : ListActivity() {
    override fun getDeleteTitle() = this.getString(R.string.account_delete_title)

    override fun getDeleteMessage(item: ListItemInterface) = this.getString(
        R.string.account_delete_message,
        if (item is Account) item.alias?.trim() else ""
    )

    override fun getListRessource() = R.layout.base_account_list_item

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val inflater = LayoutInflater.from(this)
        this.contentContainer.addView(inflater.inflate(
            R.layout.base_button_add,
            this.findViewById(android.R.id.content),
            false
        ))
        val addButton = findViewById<FloatingActionButton>(R.id.addButton)
        addButton.setOnClickListener {
            startActivityForResult(
                Intent(applicationContext, LoginActivity::class.java),
                100
            )
        }

        loadList()
    }

    override fun getId() = 0

    @SuppressWarnings("deprecation")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == 100 && resultCode == RESULT_OK) {
            loadList()
        }
    }

    override fun loadList(start: Long, limit: Long) {
        val accounts = this.application.accountModels

        if (accounts.isEmpty()) {
            this.startActivityForResult(Intent(this.applicationContext, LoginActivity::class.java), 100)
        }

        this.listAdapter.items = accounts as ArrayList<ListItemInterface>
        this.listAdapter.notifyItemRangeChanged(0, accounts.size)
    }

    override fun onClick(item: ListItemInterface) {
        if (item !is Account) {
            return
        }

        val intent = Intent(this, IndexActivity::class.java)
        AppIntentExtraService.setIntentExtra(ACCOUNT_KEY, item, intent)
//        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_MULTIPLE_TASK

        ActivityLauncherService.startActivity(
            this,
            "core",
            "desktop",
            "index",
            emptyMap(),
            this.application.getAccountById(item.id) ?: AccountDto(item)
        )
    }

    override fun bind(item: ListItemInterface, view: View) {
        if (item !is Account) {
            return
        }

        (view.findViewById<View>(R.id.alias) as TextView).text = item.alias
        (view.findViewById<View>(R.id.url) as TextView).text = item.url
        (view.findViewById<View>(R.id.user) as TextView).text = item.userName
    }

    override fun deleteItem(item: ListItemInterface): Boolean {
        if (item !is Account) {
            return false
        }

        this.runTask({
            UserTask.deleteDevice(this, item)
        })

        this.application.removeAccount(item)

        return true
    }
}